custom_components/xiaomi_miio_fan/fan.py

```diff
@@ -1748,3 +1794,320 @@ def delay_off(self, seconds: int):
+        "power": {"siid": 2, "piid": 1},
+        "fan_level": {"siid": 2, "piid": 2},
+        "oscillate": {"siid": 2, "piid": 4},
+        "angle": {"siid": 2, "piid": 5},
+        "mode": {"siid": 2, "piid": 3},
+        "delay_off_countdown": {"siid": 3, "piid": 1},
+        "child_lock": {"siid": 7, "piid": 1},
+        "light": {"siid": 4, "piid": 1},
+        "buzzer": {"siid": 5, "piid": 1},
+        "motor_control": {"siid": 6, "piid": 1},
+        "speed": {"siid": 2, "piid": 6},
+    }
+
+    def __init__(
+        self,
+        ip: str = None,
+        token: str = None,
+        start_id: int = 0,
+        debug: int = 0,
+        lazy_discover: bool = True,
+        model: str = MODEL_FAN_P33,
+    ) -> None:
+        super().__init__(ip, token, start_id, debug, lazy_discover, model=model)
+
+    def status(self):
+        """Retrieve properties."""
+        status = FanStatusP33(
+            {
+                prop["did"]: prop["value"] if prop["code"] == 0 else None
+                for prop in self.get_properties_for_mapping()
+            }
+        )
+        _LOGGER.debug("MIIO STATUS: %s", status.data)
+        return status
+
+    def on(self):
+        """Power on."""
+        return self.set_property("power", True)
+
+    def off(self):
+        """Power off."""
+        return self.set_property("power", False)
+
+    def set_speed(self, speed: int):
+        """Set fan speed."""
+        if speed < 0 or speed > 100:
+            raise FanException("Invalid speed: %s" % speed)
+
+        return self.set_property("speed", speed)
+
+    def set_angle(self, angle: int):
+        """Set the oscillation angle."""
+        if angle not in [30, 60, 90, 120, 140]:
+            raise FanException(
+                "Unsupported angle. Supported values: "
+                + ", ".join("{0}".format(i) for i in [30, 60, 90, 120, 140])
+            )
+
+        return self.set_property("angle", angle)
+
+    def set_oscillate(self, oscillate: bool):
+        """Set oscillate on/off."""
+        if oscillate:
+            return self.set_property("oscillate", True)
+        else:
+            return self.set_property("oscillate", False)
+
+    def set_buzzer(self, buzzer: bool):
+        """Set buzzer on/off."""
+        if buzzer:
+            return self.set_property("buzzer", True)
+        else:
+            return self.set_property("buzzer", False)
+
+    def set_child_lock(self, lock: bool):
+        """Set child lock on/off."""
+        self.status()
+        return self.set_property("child_lock", lock)
+
+    def set_light(self, light: bool):
+        """Set indicator state."""
+        return self.set_property("light", light)
+
+    def set_mode(self, mode: OperationModeFanP33):
+        """Set mode."""
+        # Here we convert P33 mode names (inverted) to the standard Miot ones
+        name = "Normal" if mode.name == "NaturalWind" else "Nature"
+        return self.set_property("mode", OperationModeFanZA5[name].value)
```

Comment:
Don't miss to update these lines. OperationModeP33 should be used here without the need of remappings.

Reply:
This makes so much more sense now that I look at it with fresh eye. I was convinced - for some reason - that the fan was sending strings, rather than ints -.-